pkg/filter/filter_test.go

```diff
@@ -0,0 +1,345 @@
+			TypeMeta: metav1.TypeMeta{Kind: "BuildConfig"},
+		}
+
+		mo, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&bc)
+		Expect(err).NotTo(HaveOccurred())
+
+		uo := unstructured.Unstructured{Object: mo}
+
+		err = filter.SetLabel(&uo)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(uo.GetLabels()).To(HaveKeyWithValue("specialresource.openshift.io/owned", "true"))
+	})
+})
+
+var _ = Describe("IsSpecialResource", func() {
+	cases := []struct {
+		name    string
+		obj     client.Object
+		matcher types.GomegaMatcher
+	}{
+		{
+			name: "SpecialResource",
+			obj: &v1beta1.SpecialResource{
+				TypeMeta: metav1.TypeMeta{Kind: "SpecialResource"},
+			},
+			matcher: BeTrue(),
+		},
+		{
+			name: "Pod owned by SRO",
+			obj: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"specialresource.openshift.io/owned": "true"},
+				},
+			},
+			matcher: BeFalse(),
+		},
+		{
+			name: "valid selflink",
+			obj: func() *unstructured.Unstructured {
+				uo := &unstructured.Unstructured{}
+				uo.SetSelfLink("/apis/sro.openshift.io/v1")
+
+				return uo
+			}(),
+			matcher: BeTrue(),
+		},
+		{
+			name: "selflink in Label",
+			obj: func() *unstructured.Unstructured {
+				uo := &unstructured.Unstructured{}
+				uo.SetLabels(map[string]string{"some-label": "/apis/sro.openshift.io/v1"})
+
+				return uo
+			}(),
+			matcher: BeTrue(),
+		},
+		{
+			name:    "no selflink",
+			obj:     &unstructured.Unstructured{},
+			matcher: BeFalse(),
+		},
+	}
+
+	entries := make([]TableEntry, 0, len(cases))
+
+	for _, c := range cases {
+		entries = append(entries, Entry(c.name, c.obj, c.matcher))
+	}
+
+	DescribeTable(
+		"should return the correct value",
+		func(obj client.Object, m types.GomegaMatcher) {
+			Expect(filter.IsSpecialResource(obj)).To(m)
+		},
+		entries...)
+})
+
+var _ = Describe("Owned", func() {
+	cases := []struct {
+		name    string
+		obj     client.Object
+		matcher types.GomegaMatcher
+	}{
+		{
+			name: "via ownerReferences",
+			obj: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					OwnerReferences: []metav1.OwnerReference{
+						{Kind: "SpecialResource"},
+					},
+				},
+			},
+			matcher: BeTrue(),
+		},
+		{
+			name: "via labels",
+			obj: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"specialresource.openshift.io/owned": "whatever"},
+				},
+			},
+			matcher: BeTrue(),
+		},
+		{
+			name:    "not owned",
+			obj:     &corev1.Pod{},
+			matcher: BeFalse(),
+		},
+	}
+
+	entries := make([]TableEntry, 0, len(cases))
+
+	for _, c := range cases {
+		entries = append(entries, Entry(c.name, c.obj, c.matcher))
+	}
+
+	DescribeTable(
+		"should return the expected value",
+		func(obj client.Object, m types.GomegaMatcher) {
+			Expect(filter.Owned(obj)).To(m)
+		},
+		entries...,
+	)
+})
+
+var _ = Describe("Predicate", func() {
+	resetMode := func() { filter.Mode = "" }
+
+	AfterEach(resetMode)
+
+	Context("CreateFunc", func() {
```

Comment:
I was looking at it and it looked wrong, then I understood what is the issue.

`Predicate` is an API function of the `filter` package therefore it gets it own `Describe` as you did.
But then we start calling function of an inner package of the `predicate` package in `vendor/sigs.k8s.io`.
Those inner function should be mocked and I don't think we need to test the result of those inner function (this will be `vendor/sigs.k8s.io`'s UT's scope.

Reply:
Could you clarify which inner function you are referring to?  
`CreateFunc` and others are our own code and we must test it. I don't see any better way of testing them at the moment.

Reply:
After looking at it again, there isn't much we need to change here.

The way I would expect to see it, is that if `CreateFunc` is our code, I would expect it to be tested separately in it own `Describe`.

By looking at the code, I see that the object was initialised with anonymous functions so we don't have a better way to test it.

I guess I just didn't expect the code to be written this way.

We can keep it as it is.